### PR TITLE
tsc(perf): avoid large unions

### PR DESCRIPTION
**Pull request by JonasBa**
**This is a draft PR that will not get merged, it is posted just as a proof of concept**

TLDR: we can cut TSC time almost in half and have a nicer and more responsive development experience

Longer story:
I wanted to understand why the typescript development flow on my machine was slow - code actions on save would take several seconds before type checking was done and errors due to incorrect types would pop-up, this was problematic because it decreased my development speed and resulted in multiple back and fourths between CI and TSC failing to build.

I've written down the investigation process and findings in https://www.notion.so/sentry/Typescript-performance-audit-ac375cf3d9374e66998a6eab2d5539e4

Before ~36s build time:
before
After ~21s
after

--diagnostics result before:
```
Files:              4008
Lines:            644694
Nodes:           2473488
Identifiers:      803265
Symbols:         3235184
Types:           1003633
Instantiations:  3280536
Memory used:    2772695K
I/O read:          0.46s
I/O write:         0.00s
Parse time:        2.93s
Bind time:         1.16s
Check time:       31.08s
Emit time:         0.00s
Total time:       35.16s
```
--diagnostics result after
```
Files:              4010 (+2)
Lines:            644522 (-172)
Nodes:           2468186 (-5302)
Identifiers:      801193 (-2072)
Symbols:         1627093 (-1 608 091)
Types:            540442 (-463 191)
Instantiations:  2499616 (-780 920
Memory used:    1725751K (-1 046 944K ~ 1GB)
I/O read:          0.43s (-0.03s)
I/O write:         0.00s (no change)
Parse time:        2.78s (-0.15s)
Bind time:         1.20s (+0.04s)
Check time:       17.15s (-13.93s)
Emit time:         0.00s (no change)
Total time:       21.13s (-14.03s)
```

The symbols, types, instantiations and memory usage are probably the most notable drops.

### CODE REVIEW

**Comment by evanpurkhiser** [comment]
Can't see the notion document right now, but wondering if we could split this out into a few different PRs for sets of changes?

**Comment by JonasBa** [comment]
Ah yes, sorry it was still private 🤦🏼‍♂️ - https://www.notion.so/sentry/Typescript-performance-audit-ac375cf3d9374e66998a6eab2d5539e4 I gave access to engineering

File: static/app/components/button.tsx [hunk 1]
```diff
@@ -129,7 +131,7 @@ class BaseButton extends React.Component<ButtonProps, {}> {
     // Doing this instead of using `Tooltip`'s `disabled` prop so that we can minimize snapshot nesting
     if (title) {
       return (
-        <Tooltip skipWrapper={!disabled} {...tooltipProps} title={title}>
```
Comment (scttcper) [inline static/app/components/button.tsx:132]: can we keep this, i think tooltips don't work on disabled buttons otherwise because of mouse events

Comment (JonasBa) [reply static/app/components/button.tsx:132]: @scttcper yes for sure, this branch is more of a proof of concept and we should pick whatever is useful from here. There’s a solid chance I went a bit overboard with some changes…
